### PR TITLE
WL-34 Login Functionality Backend

### PR DESCRIPTION
**Pull request by MarioRivis**
Implemented Login functionality

### CODE REVIEW

File: src/main/kotlin/ro/cadeca/weasylearn/config/KeycloakConfig.kt [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+import org.springframework.context.annotation.Bean
+import org.springframework.context.annotation.Configuration
+
+@Configuration
+class KeycloakConfig {
+    @Bean
+    fun keycloakConfigResolver(): KeycloakSpringBootConfigResolver {
+        return KeycloakSpringBootConfigResolver()
+    }
+}
```
Comment (nagyDarius) [inline src/main/kotlin/ro/cadeca/weasylearn/config/KeycloakConfig.kt]: Add new line here :)

File: src/main/kotlin/ro/cadeca/weasylearn/config/Roles.kt [hunk 2]
```diff
@@ -0,0 +1,6 @@
+package ro.cadeca.weasylearn.config
+
+const val ADMIN = "Admin"
+const val STUDENT = "Student"
+const val TEACHER = "Teacher"
+const val DEVELOPER = "Developer"
```
Comment (nagyDarius) [inline src/main/kotlin/ro/cadeca/weasylearn/config/Roles.kt]: Add new line here :)

File: src/main/kotlin/ro/cadeca/weasylearn/model/KeycloakUser.kt [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+
+class KeycloakUser(val username: String,
+                   val fullName: String,
+                   val firstName: String,
+                   val lastName: String,
+                   val roles: List<String>,
+                   val githubUsername: String? = null,
+                   val uptUsername: String? = null,
+                   val googleUsername: String? = null,
+                   val facebookUsername: String? = null)
```
Comment (nagyDarius) [inline src/main/kotlin/ro/cadeca/weasylearn/model/KeycloakUser.kt]: Add new line here :)

File: src/main/kotlin/ro/cadeca/weasylearn/services/AuthenticationService.kt [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    fun getAuthentication(): KeycloakAuthenticationToken {
+        return SecurityContextHolder.getContext().authentication as KeycloakAuthenticationToken
+    }
+
+    fun getKeycloakUser(): KeycloakUser {
+        val authentication = getAuthentication()
+        val token = authentication.account.keycloakSecurityContext.token
+        return KeycloakUser(token.preferredUsername, token.name, token.givenName, token.familyName, authentication.authorities.map { it.authority })
+    }
+}
```
Comment (nagyDarius) [inline src/main/kotlin/ro/cadeca/weasylearn/services/AuthenticationService.kt]: Add new line here :)

File: src/main/kotlin/ro/cadeca/weasylearn/config/WebSecurityConfig.kt [hunk 5]
```diff
@@ -1,14 +1,48 @@
...
+        http.cors()
         http.csrf().disable()
     }
+
+    @Autowired
+    fun configureGlobal(auth: AuthenticationManagerBuilder?) {
+        super.configure(auth)
+        val keycloakAuthenticationProvider = keycloakAuthenticationProvider()
+        keycloakAuthenticationProvider.setGrantedAuthoritiesMapper(SimpleAuthorityMapper())
+        auth!!.authenticationProvider(keycloakAuthenticationProvider)
```
Comment (nagyDarius) [inline src/main/kotlin/ro/cadeca/weasylearn/config/WebSecurityConfig.kt]: This would look better as a one liner

**Review by nagyDarius — CHANGES_REQUESTED** [review]

File: src/main/kotlin/ro/cadeca/weasylearn/config/WebSecurityConfig.kt (hunk 5 above)
Comment (MarioRivis) [reply src/main/kotlin/ro/cadeca/weasylearn/config/WebSecurityConfig.kt]: too clumsy to understand if just one line

File: src/main/kotlin/ro/cadeca/weasylearn/services/AuthenticationService.kt (hunk 4 above)
Comment (MarioRivis) [reply src/main/kotlin/ro/cadeca/weasylearn/services/AuthenticationService.kt]: Could we maybe add this to IntelliJ on autoformat?

**Review by nagyDarius — APPROVED** [review]

**Pull request merged**